// @subsquid/substrate-metadata/lib/old/definitions/metadata/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.types = void 0;
const base_1 = require("./base");
const v10_1 = require("./v10");
const v11_1 = require("./v11");
const v12_1 = require("./v12");
const v13_1 = require("./v13");
const v14_1 = require("./v14");
const v9_1 = require("./v9");
exports.types = {
    Type: 'Str',
    ...base_1.ScaleInfoTypes,
    ...v9_1.V9,
    ...v10_1.V10,
    ...v11_1.V11,
    ...v12_1.V12,
    ...v13_1.V13,
    ...v14_1.V14,
    Metadata: {
        _enum: {
            V0: 'DoNotConstruct',
            V1: 'DoNotConstruct',
            V2: 'DoNotConstruct',
            V3: 'DoNotConstruct',
            V4: 'DoNotConstruct',
            V5: 'DoNotConstruct',
            V6: 'DoNotConstruct',
            V7: 'DoNotConstruct',
            V8: 'DoNotConstruct',
            // First version on Kusama is V9, dropping will be problematic
            V9: 'MetadataV9',
            V10: 'MetadataV10',
            V11: 'MetadataV11',
            V12: 'MetadataV12',
            V13: 'MetadataV13',
            V14: 'MetadataV14'
        }
    }
};
