// @subsquid/substrate-metadata/lib/old/definitions/kilt/types_21.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.types21 = void 0;
const types_20_1 = require("./types_20");
exports.types21 = {
    ...types_20_1.types20,
    StorageError: {
        _enum: {
            DidAlreadyPresent: 'Null',
            DidNotPresent: 'Null',
            DidKeyNotPresent: 'DidVerificationKeyRelationship',
            VerificationKeyNotPresent: 'Null',
            CurrentlyActiveKey: 'Null',
            MaxTxCounterValue: 'Null',
            MaxPublicKeysPerDidKeyIdentifierExceeded: 'Null',
            MaxTotalKeyAgreementKeysExceeded: 'Null',
            MaxOldAttestationKeysExceeded: 'Null'
        }
    },
    DidCreationDetails: {
        did: 'DidIdentifierOf',
        newKeyAgreementKeys: 'DidNewKeyAgreementKeys',
        newAssertionMethodKey: 'Option<DidVerificationKey>',
        newCapabilityDelegationKey: 'Option<DidVerificationKey>',
        newServiceEndpoints: 'Option<ServiceEndpoints>'
    },
    DidDetails: {
        authenticationKey: 'KeyIdOf',
        keyAgreementKeys: 'DidKeyAgreementKeys',
        capabilityDelegationKey: 'Option<KeyIdOf>',
        assertionMethodKey: 'Option<KeyIdOf>',
        publicKeys: 'DidPublicKeyMap',
        serviceEndpoints: 'Option<ServiceEndpoints>',
        lastTxCounter: 'u64'
    },
    DelegateSignatureTypeOf: 'DidSignature',
    ContentType: {
        _enum: ['application/json', 'application/ld+json']
    },
    Collator: {
        id: 'AccountId',
        stake: 'Balance',
        delegators: 'Vec<Stake>',
        total: 'Balance',
        state: 'CollatorStatus'
    },
    Delegator: {
        delegations: 'Vec<Stake>',
        total: 'Balance'
    },
    Keys: 'SessionKeys1'
};
