diff --git a/node_modules/@subsquid/typeorm-codegen/lib/codegen.js b/node_modules/@subsquid/typeorm-codegen/lib/codegen.js
index 54b1dd0..59c5437 100644
--- a/node_modules/@subsquid/typeorm-codegen/lib/codegen.js
+++ b/node_modules/@subsquid/typeorm-codegen/lib/codegen.js
@@ -124,7 +124,7 @@ function generateOrmModels(model, dir) {
                         break;
                     case 'lookup':
                         imports.useTypeorm('OneToOne');
-                        out.line(`@OneToOne_(() => ${prop.type.entity})`);
+                        out.line(`@OneToOne_(() => ${prop.type.entity}, e => e.${prop.type.field})`);
                         break;
                     case 'list-lookup':
                         imports.useTypeorm('OneToMany');
