// @subsquid/util-internal-config/lib/parse.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonSyntaxError = exports.parse = void 0;
const jsonc = __importStar(require("jsonc-parser"));
function parse(text) {
    let errors = [];
    let obj = jsonc.parse(text, errors, {
        allowTrailingComma: true,
        allowEmptyContent: false
    });
    if (errors.length > 0) {
        let err = errors[0];
        throw new JsonSyntaxError(jsonc.printParseErrorCode(err.error), err.offset, err.length);
    }
    return obj;
}
exports.parse = parse;
class JsonSyntaxError extends Error {
    constructor(code, offset, length) {
        super();
        this.code = code;
        this.offset = offset;
        this.length = length;
    }
}
exports.JsonSyntaxError = JsonSyntaxError;
