// @subsquid/util-xxhash/lib/composite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XXHash256 = exports.XXHash128 = void 0;
class XXHash128 {
    constructor(h64) {
        this.h0 = h64(0);
        this.h1 = h64(1);
    }
    update(data) {
        this.h0.update(data);
        this.h1.update(data);
        return this;
    }
    digest() {
        let d1 = this.h0.digest();
        let d2 = this.h1.digest();
        return d1 + (d2 << 64n);
    }
}
exports.XXHash128 = XXHash128;
class XXHash256 {
    constructor(h64) {
        this.h0 = h64(0);
        this.h1 = h64(1);
        this.h2 = h64(2);
        this.h3 = h64(3);
    }
    update(data) {
        this.h0.update(data);
        this.h1.update(data);
        this.h2.update(data);
        this.h3.update(data);
        return this;
    }
    digest() {
        let d0 = this.h0.digest();
        let d1 = this.h1.digest();
        let d2 = this.h2.digest();
        let d3 = this.h3.digest();
        return d0 + (d1 << 64n) + (d2 << 128n) + (d3 << 192n);
    }
}
exports.XXHash256 = XXHash256;
