// @subsquid/substrate-metadata-explorer/lib/out.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Out = void 0;
const util_internal_1 = require("@subsquid/util-internal");
const fs = __importStar(require("fs"));
const path_1 = require("path");
const specVersion_1 = require("./specVersion");
class Out {
    constructor(file) {
        this.file = file;
    }
    exists() {
        return fs.existsSync(this.file);
    }
    isJson() {
        return (0, path_1.extname)(this.file) == '.json';
    }
    knownVersions() {
        if (this.exists()) {
            return (0, specVersion_1.readSpecVersions)(this.file);
        }
        else {
            return [];
        }
    }
    append(versions) {
        if (this.isJson()) {
            let current = this.exists() ? (0, specVersion_1.readSpecVersions)(this.file) : [];
            this.write(current.concat(versions));
        }
        else {
            for (let v of versions) {
                fs.appendFileSync(this.file, JSON.stringify(v) + '\n');
            }
        }
    }
    write(versions) {
        if (this.isJson()) {
            fs.writeFileSync(this.file, JSON.stringify(versions, null, 2));
        }
        else {
            fs.unlinkSync(this.file);
            for (let v of versions) {
                fs.appendFileSync(this.file, JSON.stringify(v) + '\n');
            }
        }
    }
}
__decorate([
    util_internal_1.def,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Array)
], Out.prototype, "knownVersions", null);
exports.Out = Out;
