// @subsquid/substrate-metadata-explorer/lib/util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.withErrorContext = exports.addErrorContext = exports.printValidationErrors = exports.makeValidator = void 0;
const ajv_1 = __importDefault(require("ajv"));
function makeValidator(schema) {
    let ajv = new ajv_1.default({
        messages: true,
        removeAdditional: false,
        verbose: true
    });
    return ajv.compile(schema);
}
exports.makeValidator = makeValidator;
function printValidationErrors(validator, separator = ', ') {
    if (!validator.errors?.length)
        return '';
    return validator.errors.map((e) => `data${e.instancePath} ${e.message}`)
        .reduce((text, msg) => text + separator + msg);
}
exports.printValidationErrors = printValidationErrors;
function addErrorContext(err, ctx) {
    let e = err;
    for (let key in ctx) {
        switch (key) {
            case 'blockHeight':
            case 'blockHash':
                if (e.blockHeight == null && e.blockHash == null) {
                    e.blockHeight = ctx.blockHeight;
                    e.blockHash = ctx.blockHash;
                }
                break;
            default:
                if (e[key] == null) {
                    e[key] = ctx[key];
                }
        }
    }
    return err;
}
exports.addErrorContext = addErrorContext;
function withErrorContext(ctx) {
    return function (err) {
        throw addErrorContext(err, ctx);
    };
}
exports.withErrorContext = withErrorContext;
